packages/steps/src/utils/shell/spawn.ts

```diff
@@ -1,45 +1,161 @@
-  if (logger && promise.child) {
-    pipeSpawnOutput(logger, promise.child, options);
+): Promise<SpawnResult> {
+  const { logger, noLogsTimeout, ...options } = allOptions;
+  let spawnWarnTimeout: NodeJS.Timeout | undefined;
+  let spawnKillTimeout: NodeJS.Timeout | undefined;
+  let spawnTimedOut: boolean = false;
+  let spawnResult: SpawnResult;
+
+  function setCommandSpawnTimeouts(
+    noLogsTimeout: NoLogsTimeoutOptions,
+    logger: bunyan,
+    spawnPromise: SpawnPromise<SpawnResult>
+  ): void {
+    if (noLogsTimeout.warn) {
+      const warnTimeoutMinutes =
+        noLogsTimeout.warn.timeoutMinutes ?? SPAWN_WARN_TIMEOUT_DEFAULT_MINUTES;
+      spawnWarnTimeout = setTimeout(
+        () => {
+          logger.warn(
+            noLogsTimeout.warn?.message ??
+              SPAWN_WARN_TIMEOUT_DEFAULT_MESSAGE.replace(
+                '${minutes}',
+                warnTimeoutMinutes.toString()
+              )
+          );
+        },
+        warnTimeoutMinutes * 60 * 1000
+      );
+    }
+
+    if (noLogsTimeout.kill) {
+      const killTimeoutMinutes =
+        noLogsTimeout.kill.timeoutMinutes ?? SPAWN_KILL_TIMEOUT_DEFAULT_MINUTES;
+      spawnKillTimeout = setTimeout(
+        async () => {
+          spawnTimedOut = true;
+          logger.error(
+            noLogsTimeout.kill?.message ??
+              SPAWN_KILL_TIMEOUT_DEFAULT_MESSAGE.replace(
+                '${minutes}',
+                killTimeoutMinutes.toString()
+              )
+          );
+          const ppid = nullthrows(spawnPromise.child.pid);
+          const pids = await getParentAndDescendantProcessPidsAsync(ppid);
+          pids.forEach((pid) => {
+            process.kill(pid);
+          });
+        },
+        killTimeoutMinutes * 60 * 1000
+      );
+    }
+  }
+
+  try {
+    const spawnPromise = spawnAsyncOriginal(command, args, options);
+    if (logger && spawnPromise.child) {
+      if (noLogsTimeout) {
+        const optionsWithCallback = {
+          ...options,
+          infoCallbackFn: () => {
+            if (spawnWarnTimeout) {
+              spawnWarnTimeout.refresh();
+            }
+            if (spawnKillTimeout) {
+              spawnKillTimeout.refresh();
+            }
+          },
+        };
+        pipeSpawnOutput(logger, spawnPromise.child, optionsWithCallback);
+        setCommandSpawnTimeouts(noLogsTimeout, logger, spawnPromise);
+      } else {
+        pipeSpawnOutput(logger, spawnPromise.child, options);
+      }
+    }
+    spawnResult = await spawnPromise;
+  } catch (err: any) {
+    if (spawnTimedOut) {
+      const spawnKillTimeout =
+        noLogsTimeout?.kill?.timeoutMinutes ?? SPAWN_KILL_TIMEOUT_DEFAULT_MINUTES;
+      const errorMessage =
+        noLogsTimeout?.kill?.errorMessage ??
+        SPAWN_KILL_TIMEOUT_DEFAULT_ERROR_MESSAGE.replace('${minutes}', spawnKillTimeout.toString());
+      const ErrorClass = noLogsTimeout?.kill?.errorClass ?? Error;
```

Comment:
Can we use some well-known error class for this? Maybe we want to use `UserFacingError` for this, or add a new subclass of `UserFacingError` in `eas-build-job` and use that?